### ISSUE TITLE
[Feature Request] Azure-repository snapshot support Client-side encryption for blobs

### ISSUE THREAD

**Issue by cthtrifork** [opensearch-project/OpenSearch#14222]
### Is your feature request related to a problem? Please describe

We would like the plugin to support client side encryption:
https://learn.microsoft.com/en-us/azure/storage/blobs/client-side-encryption?tabs=java

Is there support for adopting this feature? We might make the pull request but I wanted to hear the opinion of the maintainers.

### Describe the solution you'd like

Support the optional setup for client side encryption and allow configuration of access to Azure Key Vault.

### Related component

Plugins

### Additional context

The Azure Blob Storage client libraries use envelope encryption to encrypt and decrypt your data on the client side. Envelope encryption encrypts a key with one or more additional keys.

The Blob Storage client libraries rely on Azure Key Vault to protect the keys that are used for client-side encryption. For more information about Azure Key Vault, see [What is Azure Key Vault?](https://learn.microsoft.com/en-us/azure/key-vault/general/overview).

**Comment by cthtrifork** [comment]
I think we should also look into support managed identities for Azure Key Vault as well. As it became supported in https://github.com/opensearch-project/OpenSearch/issues/12423

**Comment by andrross** [comment]
[Triage - attendees [1](https://github.com/scrawfor99) [2](https://github.com/reta) [3](https://github.com/andrross) [4](https://github.com/jed326/) [5](https://github.com/srikanthpadakanti/)]

Thanks for filing. We do have work in flight for general client-side encryption for all repository stores: #5800, #7229

We could also explore the option of adding the option to use the feature within repository-azure as well as an alternative. I'm not totally opposed to that, but would be happy to hear the opinion from others as well.

/cc @vikasvb90

**Comment by vikasvb90** [comment]
@cthtrifork We rolled out client side encryption support outside of the repository plugins on top of repository layer so that encryption and repository plugins remain decoupled. 
The whole framework is divided in two parts:
1. **Crypto Key provider plugin**: A plugin of this [type ](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/CryptoKeyProviderPlugin.java) is responsible for providing data key pairs to be used in encryption/decryption. There is a AWS KMS based [crypto-kms](https://github.com/opensearch-project/OpenSearch/tree/main/plugins/crypto-kms) plugin available which retrieves keys from AWS KMS service. Similar plugins can be added for other master key providers.
2. **Encryption/Decryption layer**: A crypto abstraction plugin is available in OpenSearch [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/CryptoPlugin.java) to encrypt or decrypt provided streams using the configured key provider against a repository. Unfortunately, there is no default plugin available for this yet and hence, you may need to build and keep a custom implementation. Once this plugin or module is available, this gets loaded by [registry](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/crypto/CryptoHandlerRegistry.java#L152-L153) against the configured encrypted repository. 

Once, both of these components are available, you can create an encrypted repository. Following is how a sample request would look like.

```
PUT _snapshot/vikasvb-repository-test-1
{
  "type": "s3",
  "settings": { // Repository settings
    "bucket": "<bucket-name>",
    "region": "<region>"
  },    
  "crypto_settings": {
      "key_provider_name": "first-mock-1",
      "key_provider_type": "sample-key-provider-extension", // Key Provider Type
      "settings": { // Key provider settings 
          "key-1": "sample value"
      }
  }
}
```

`key_provider_type` here is the identifier you will configure for your key provider impl. `aws-kms` is the type for the crypto-kms plugin as mentioned [here](https://github.com/opensearch-project/OpenSearch/blob/main/plugins/crypto-kms/src/main/java/org/opensearch/crypto/kms/CryptoKmsPlugin.java#L27).

**Comment by cthtrifork** [comment]
@vikasvb90 That is great. I don't think it make sense to do a proprietary/vendor solution if there is upcoming support for a generic approach. Any thoughts  on when this could be available? Summer? Fall? Winter?

**Comment by linuxpi** [comment]
[Storage Triage - attendees [1](https://github.com/ashking94/) [2](https://github.com/shourya035/) [3](https://github.com/gbbafna/) [4](https://github.com/andrross/) [5](https://github.com/ltaragi/) [6](https://github.com/astute-decipher/) [7](https://github.com/skumawat2025/) [8](https://github.com/linuxpi/)]

@vikasvb90 Please share some details if you have any on which version can we expect to see this!

**Comment by cthtrifork** [comment]
@andrross Do you have any more insights to share? any news? To my understanding, we are waiting for https://github.com/opensearch-project/OpenSearch/issues/5800

**Comment by cthtrifork** [comment]
To be clear, I am hoping clientside encryption becomes supported in `repository-azure`. It would make a lot of sense to use core capabilities instead of relying on a azure specific solution, if there has been a lot of progress on core support already.

**Comment by vikasvb90** [comment]
@cthtrifork As I mentioned earlier, if an implementation of encryption plugin is available, core already has capability to register encrypted snapshots. Entire thing is decoupled and you don't need to put encryption within a repository plugin. Regarding impl of encryption plugin, there's no plan yet but feel free to refer this old PR https://github.com/opensearch-project/OpenSearch/pull/9823. 

Please refer to my previous comment on how an implementation of encryption plugin can allow registration of encrypted snapshots automatically.

**Comment by cthtrifork** [comment]
@vikasvb90 Thanks a lot for getting back to me. Perhaps I am just missing some understanding of the concept. 

So for my understanding. If we have this configured:

PUT _snapshot/vikasvb-repository-test-1
{
  "type": "s3",
  "settings": { // Repository settings
    "bucket": "<bucket-name>",
    "region": "<region>"
  },    
  "crypto_settings": {
      "key_provider_name": "user-db-repository-key",
      "key_provider_type": "aws-kms", // Key Provider Type
      "settings": { // Key provider settings 
          "key-1": "sample value"
      }
  }
}

It would work today? given that we have a `aws-kms` setup in AWS and have loaded the `crypto-kms` plugin.

In our case, we would need to make a new plugin for the KMS part, as we can not rely on AWS-KMS, but have to make a plugin for Azure vault or hashicorp vault then.

**Comment by vikasvb90** [comment]
Correct! you will need to build it for azure. But in addition to this you also need to implement the actual encryption interface in the form of a module. You can refer to the old PR which is a bit complex https://github.com/opensearch-project/OpenSearch/pull/9823 but you can choose to provide a simpler or different encryption implementation.

**Comment by cthtrifork** [comment]
Thanks so much for clarifying. It seems like there is a lot of work left and no official work being done to progress this at the moment. 

I think we will try and test if https://github.com/Aiven-Open/encrypted-repository-opensearch covers our needs instead, as I am hesitant to implement an encryption implementation in a safe, performant and secure manner.
